### PR TITLE
allow dots in config set

### PR DESCRIPTION
**Pull request by deads2k**
`kubectl config set` did not allow for dots in the names of clusters, contexts, or users.  This is a problem if you're autogenerating keys that have IP addresses in them.  This allows dots by greedily consuming extra tokens until an expected key is found.  

/cc @smarterclayton @jlowdermilk

### CODE REVIEW

File: pkg/kubectl/cmd/config/navigation_step_parser.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+			if !configValueOptions.Has(nextPart) {
+				return nil, fmt.Errorf("unable to parse %v after %v at %v", path, steps, currType)
+			}
+
+			steps = append(steps, navigationStep{nextPart, nextType})
+			currPartIndex += 1
+			currType = nextType
+
+		case cluster:
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/config/navigation_step_parser.go]: yuck - anyway to dry this up?

File: pkg/kubectl/cmd/config/navigation_step_parser.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	preferenceValue = stepLocation("preference-value")
+	cluster         = stepLocation("cluster")
+	clusterName     = stepLocation("cluster-name")
+	clusterValue    = stepLocation("cluster-value")
+	context         = stepLocation("context")
+	contextName     = stepLocation("context-name")
+	contextValue    = stepLocation("context-value")
+	authInfo        = stepLocation("authInfo")
+	authInfoName    = stepLocation("authInfo-name")
+	authInfoValue   = stepLocation("authInfo-value")
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/config/navigation_step_parser.go]: How do I as a developer know when to add something to this list?

File: pkg/kubectl/cmd/config/navigation_step_parser.go (hunk 1 above)
Comment (deads2k) [inline pkg/kubectl/cmd/config/navigation_step_parser.go]: > yuck - anyway to dry this up?

Metadata map or reflection (blech).  Since the metadata can be built using reflection and that addresses your other concern, I'll do it like that.

File: pkg/util/set.go [hunk 3]
```diff
@@ -32,6 +33,18 @@ func NewStringSet(items ...string) StringSet {
 	return ss
 }
 
+// KeySet creates a StringSet from a keys of a map[string](? extends interface{}).  Since you can't describe that map type in the Go type system
+// the reflected value is required.
+func KeySet(theMap reflect.Value) StringSet {
```
Comment (deads2k) [inline pkg/util/set.go]: @liggitt  Comments here?  I know you've faced the problem before.

Comment (liggitt) [inline pkg/util/set.go]: I'd never tried to do it on a generic map type... can't really think of anything better, but ick...

**Comment by deads2k** [comment]
@smarterclayton switched to reflective implementation

**Comment by deads2k** [comment]
@jlowdermilk any comments?

File: pkg/kubectl/cmd/config/navigation_step_parser.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+	individualParts := strings.Split(path, ".")
+
+	currType := reflect.TypeOf(clientcmdapi.Config{})
+	currPartIndex := 0
+	for currPartIndex < len(individualParts) {
+		switch currType.Kind() {
+		case reflect.Map:
+			// if we're in a map, we need to locate a name.  That name may contain dots, so we need to know what tokens are legal for the map's value type
+			mapValueType := currType.Elem()
+			mapValueOptions := getPotentialTypeValues(mapValueType)
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/config/navigation_step_parser.go]: The logic here is quite hard to follow and deserves a longer comment explaining what's going on.

File: pkg/kubectl/cmd/config/navigation_step_parser.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+	}
+	nextParts := parts[0:numberOfPartsInStep]
+
+	return strings.Join(nextParts, ".")
+}
+
+func getPotentialTypeValues(typeValue reflect.Type) map[string]reflect.Type {
+	ret := make(map[string]reflect.Type)
+
+	for fieldIndex := 0; fieldIndex < typeValue.NumField(); fieldIndex++ {
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/config/navigation_step_parser.go]: NumField will panic if typeValue.Kind() is not struct. I don't grok this enough to be sure that won't happen.

File: pkg/kubectl/cmd/config/navigation_step_parser.go (hunk 4 above)
Comment (deads2k) [inline pkg/kubectl/cmd/config/navigation_step_parser.go]: > The logic here is quite hard to follow and deserves a longer comment explaining what's going on.

Comment added.

**Comment by jlowdermilk** [comment]
LGTM, thanks for the added comments.

File: pkg/kubectl/cmd/config/navigation_step_parser.go (hunk 5 above)
Comment (deads2k) [inline pkg/kubectl/cmd/config/navigation_step_parser.go]: > NumField will panic if typeValue.Kind() is not struct. I don't grok this enough to be sure that won't happen.

It shouldn't happen, but I've added a guard to prevent the panic.
